docs/proposals/202001_thanos_query_health_handling.md

```diff
@@ -0,0 +1,86 @@
+
+Thus, this logic needs to be changed somehow. There are a few possible options:
+
+* `--store.unhealthy-timeout` could be made to apply to this case as well - we could still consider it a part of the active store set while it is still visible in the UI
+* Another option could be introduced such as `--store.hold-timeout` which would be `--store.unhealthy-timeout`'s brother and we would hold the StoreAPI nodes for `max(hold_timeout, unhealthy_timeout)`.
+* Another option such as `--store.strict-mode` could be introduced which means that we would always retain the last information of the StoreAPI nodes after the first successful check
+* The StoreAPI node specification format that is used in `--store` could be extended to include another flag which would let specify the previous option per-specific node
+
+Lets look through their pros and cons:
+
+* In general it would be nice to avoid adding new options so the first option seems the most attractive in this regard but it is also the most invasive one;
+* Second option increases the configuration complexity the most since you would have to think about the other option `--store.unhealthy-timeout` as well while setting it;
+* The last option is the most complex from code's perspective but it is also least invasive; however it has some downsides like the syntax for specifying store nodes becomes really ugly and hard to understand because we would have not only the DNS query type in there but also a special marker to enable that mode
+
+If we were to graph these choices in terms of theirs incisiveness and complexity it would look something like this:
+
+```text
+Most incisive / Least Complex ------------ Least incisive / Most Complex
+#1           #2                                      #4
+           #3
+```
+
+At first there was an initial attempt to add the last option but it was met with some discussion if this is really the way to go forward. After careful consideration and with the rationale in this proposal we could go with the most invasive choice. But it also probably makes the most sense since it would logically follow that as long as a node is visible in the UI then we send a query to it.
+
+## Goals
+
+* Update the health check logic to properly handle the case when a node disappears in terms of a caching layer
+
+## No Goals
+
+* Fixing the cache handling in the cases where a **new** StoreAPI gets added to the active store set
+
+## Verification
+
+* Unit tests which would fire up a dummy StoreAPI and check different scenarios
+* Ad-hoc testing
+
+## Proposal
+
+* Keep store nodes around in the active store set in Thanos Query until `--store.unhealthy-timeout` passes
+
+## Risk
+
+* Users might start having partial responses with this implemented if they have certain nodes configured that are always down. However, one might argue that if the nodes go down then something like DNS service discovery needs to be used which would dynamically add and remove those nodes
```

Comment:
trailing period but also I don't really understand this sentence, can you rephrase?: 

>  However, one might argue that if the nodes go down then something like DNS service discovery needs to be used which would dynamically add and remove those nodes

Reply:
Rephrased it with your suggestion + some extra text. Let me know if it makes sense.